zstd.cc

```diff
@@ -0,0 +1,134 @@
+/*
+ * Copyright (C) 2019 ScyllaDB
+ */
+
+/*
+ * This file is part of Scylla.
+ *
+ * Scylla is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * Scylla is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <seastar/core/aligned_buffer.hh>
+
+// We need to use experimental features of the zstd library (to allocate compression/decompression context),
+// which are available only when the library is linked statically.
+#define ZSTD_STATIC_LINKING_ONLY
+#include <zstd.h>
+
+#include "compress.hh"
+#include "utils/class_registrator.hh"
+
+static const sstring COMPRESSION_LEVEL = "compression_level";
+static const sstring COMPRESSOR_NAME = compressor::namespace_prefix + "ZstdCompressor";
+
+class zstd_processor : public compressor {
+    int _compression_level = 3;
+
+    // Compression context.
+    std::unique_ptr<char[], free_deleter> _cctx_raw;
+    ZSTD_CCtx* _cctx;
+
+    // Decompression context.
+    std::unique_ptr<char[], free_deleter> _dctx_raw;
+    ZSTD_DCtx* _dctx;
```

Comment:
I see initialization code for this pointer, but no de-initialization (the class doesn't even have a destructor).

Reply:
The default destructor does the right thing (i.e., calling the destructor of these unique_ptrs, which free the memory allocated for the contexts)
https://github.com/facebook/zstd/blob/dev/lib/zstd.h#L1272

Reply:
I see. Maybe `_dctx_storage` would be a better name then (and consequently `_cctx_storage`).

Reply:
What about `_dctx_bytes`?

Reply:
`bytes` sounds more like a unit of measurement to me (for size), than a storage place. I guess raw is fine too, maybe add a comment.

Reply:
Added a comment